swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Data.qll

```diff
@@ -1,6 +1,47 @@
 import swift
 private import codeql.swift.dataflow.ExternalFlow
 
+private class DataSources extends SourceModelCsv {
+  override predicate row(string row) {
+    row = ";Data;true;init(contentsOf:options:);;;ReturnValue;remote"
+  }
+}
+
 private class DataSummaries extends SummaryModelCsv {
-  override predicate row(string row) { row = ";Data;true;init(_:);;;Argument[0];ReturnValue;taint" }
+  override predicate row(string row) {
+    row =
+      [
+        ";Data;true;init(_:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(base64Encoded:options:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(buffer:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(bytes:count:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(contentsOf:options:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(bytesNoCopy:count:deallocator:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;init(referencing:);;;Argument[0];ReturnValue;taint",
+        ";Data;true;append(_:);;;Argument[0];Argument[-1];taint",
+        ";Data;true;append(_:count:);;;Argument[0];Argument[-1];taint",
+        ";Data;true;append(contentsOf:);;;Argument[0];Argument[-1];taint",
+        ";Data;true;base64EncodedData(options:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;base64EncodedString(options:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;compactMap(_:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;copyBytes(to:);;;Argument[-1];Argument[0];taint",
+        ";Data;true;copyBytes(to:count:);;;Argument[-1];Argument[0];taint",
+        ";Data;true;copyBytes(to:from:);;;Argument[-1];Argument[0];taint",
+        ";Data;true;flatMap(_:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;insert(_:at:);;;Argument[0];Argument[-1];taint",
+        ";Data;true;insert(contentsOf:at:);;;Argument[0];Argument[-1];taint",
+        ";Data;true;map(_:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;reduce(into:_:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;replace(_:with:maxReplacements:);;;Argument[1];Argument[-1];taint",
+        ";Data;true;replaceSubrange(_:with:);;;Argument[1];Argument[-1];taint",
+        ";Data;true;replaceSubrange(_:with:count:);;;Argument[1];Argument[-1];taint",
+        ";Data;true;replacing(_:with:maxReplacements:);;;Argument[1];Argument[-1];taint",
+        ";Data;true;replacing(_:with:subrange:maxReplacements:);;;Argument[1];Argument[-1];taint",
+        ";Data;true;shuffled();;;Argument[-1];ReturnValue;taint",
+        ";Data;true;shuffled(using:);;;Argument[-1];ReturnValue;taint",
+        ";Data;true;sorted(using:);;;Argument[-1];ReturnValue;taint",
```

Comment:
We could perhaps add [`sorted()`](https://developer.apple.com/documentation/foundation/data/1779722-sorted), [`sorted(by:)`](https://developer.apple.com/documentation/foundation/data/2293440-sorted) and [`reversed()`](https://developer.apple.com/documentation/foundation/data/1780245-reversed)?

Reply:
Done, except `reversed()`. That got me banging my head against the wall for a while, but in the end the thing is that the `reverse()` implementation actually comes from an extension of the `BidirectionalCollection` protocol. Which incidentally makes me think we need models around that and, more generally, `Collection`.

We could always create a stub `reverse` function in `Data` that forces the return object to be `ReversedCollection<Data>` with some weird cast (because we can't realistically instantiate that type), but I'm not sure that's actually useful if in real databases `Data` doesn't actually override `reverse()`.

For the moment I added a `MISSING` test case for that. but let me know if you prefer the latter solution.

Reply:
Thanks for investigating and adding the two `sorted` methods.

I don't think `reverse` is all the important as an end in itself.

You're probably right about `Collection`.  There are a whole load of standard collections we haven't even thought about modelling yet, some will be more important than others.  I think I'll create an issue just to discuss and prioritize this.